include/rocksdb/comparator.h

```diff
@@ -79,6 +79,12 @@ class Comparator {
   // The major use case is to determine if DataBlockHashIndex is compatible
   // with the customized comparator.
   virtual bool CanKeysWithDifferentByteContentsBeEqual() const { return true; }
+
+  virtual size_t TimestampSize() const { return 0; }
```

Comment:
I believe having all the timestamps to be of the same length would be very tight restriction for the users to adopt this feature. Can this be relaxed?
Example: The timestamp field could be used by users to specify "version numbers". Lets say the version numbers start from 1 and monotonically increase. Since we are now using a slice to represent the timestamps, the length of the version numbers would keep varying as they increase, unless the users know ahead of time to start from an arbitrarily large number which never increases by an order of magnitude.

Reply:
> Once created, the format and length of the timestamp cannot change (at least for now).

Any tentative plan to relax this in the future?

Reply:
> I believe having all the timestamps to be of the same length would be very tight restriction for the users to adopt this feature. Can this be relaxed?
> Example: The timestamp field could be used by users to specify "version numbers". Lets say the version numbers start from 1 and monotonically increase. Since we are now using a slice to represent the timestamps, the length of the version numbers would keep varying as they increase, unless the users know ahead of time to start from an arbitrarily large number which never increases by an order of magnitude.

I think the user can use a `uint64_t` (fixed-length slice) to represent the version number in this case rather than encoding the version numbers into variable-length slices. I have been thinking of other cases in which users use Lamport time, logical clock, NTP or HLC with a well-defined, widely-agreed-upon standard on the timestamp format. In these cases, timestamps are of fixed-length, and do not change very often.

Reply:
> > Once created, the format and length of the timestamp cannot change (at least for now).
> 
> Any tentative plan to relax this in the future?

One approach I can think of now, is to use some technique similar to compaction filter and perform a full compaction.

Reply:
I'm concerned about the performance penalty of a varlen encoding/decoding here. I believe fixed length timestamp is dominating anyway.